components/license-control-service/pkg/storage/storage.go

```diff
@@ -0,0 +1,248 @@
+// have a configured License.
+type NoLicenseError struct{ backend string }
+
+func (n *NoLicenseError) Error() string {
+	return fmt.Sprintf("No license found in %s storage backend", n.backend)
+}
+
+// RetriableBackendError is the error returned by a backend when an
+// update has failed but may work if retried.
+type RetriableBackendError struct{ err error }
+
+func (r *RetriableBackendError) Error() string { return r.err.Error() }
+
+// A CurrentBackend can be used as a storage backend for the
+// license-control-service.
+type CurrentBackend interface {
+	// Init prepares the backend and migrates data from any known
+	// Upgradeable Backends. Must be called before GetLicense and
+	// SetLicense.
+	Init(context.Context, *keys.LicenseParser) error
+	// GetLicense returns the currently configured autoamte
+	// license. If no license is configured, NoLicense is
+	// returend.
+	GetLicense(context.Context) (string, keys.LicenseMetadata, error)
+	// SetLicense stores the given license in the backend.
+	SetLicense(context.Context, string) error
+}
+
+// An UpgradeableBackend is an old backend that can no longer be used
+// as a backend for the license-control-service but which might have
+// data that needs to be migrated into the CurrentBackend.
+type UpgradeableBackend interface {
+	// GetLicense returns the currently configured autoamte
+	// license. If no license is configured, ErrNoLicense is
+	// returend.
+	GetLicense(context.Context) (string, error)
+	// Cleanup removes the state kept by this backend. After
+	// Cleanup is called, GetLicense should return a
+	// NoLicenseError.
+	Cleanup(context.Context) error
+}
+
+var _ CurrentBackend = (*PGBackend)(nil)
+var _ UpgradeableBackend = (*FileBackend)(nil)
+
+func NewCurrentBackend(pgURL, migrationPath, legacyFilePath string) CurrentBackend {
+	return &PGBackend{
+		pgURL:             pgURL,
+		migrationPath:     migrationPath,
+		legacyFileBackend: &FileBackend{path: legacyFilePath},
+	}
+}
+
+// PgBackend is the CurrentBackend used in production. It stores the
+// raw key data in PostgreSQL.
+type PGBackend struct {
+	pgURL             string
+	migrationPath     string
+	legacyFileBackend UpgradeableBackend
+
+	db *sql.DB
+}
+
+func (p *PGBackend) Init(ctx context.Context, l *keys.LicenseParser) error {
+	db, err := sql.Open("postgres", p.pgURL)
+	if err != nil {
+		return errors.Wrap(err, "failed to open database connection")
+	}
+	p.db = db
+
+	err = migrator.Migrate(p.pgURL, p.migrationPath, logger.NewLogrusStandardLogger(), false)
+	if err != nil {
+		return errors.Wrap(err, "failed to apply database schema")
+	}
+
+	_, _, err = p.GetLicense(ctx)
+	switch err.(type) {
+	case nil:
+		// Already migrated
+		return nil
+	case *NoLicenseError:
+		return p.migrateFromFileBackend(ctx, l)
+	default:
+		return err
+	}
+}
+
+// TODO(ssd) 2019-07-29: This migration would be problematic in
+// multi-node, but this should only happen on single node clusters.
```

Comment:
why? I assume it's because to migrate to HA you'll have to do some other things that would make this upgrade occur on a single-node, but would be nice to know the reasoning.